modules/renter/download.go

```diff
@@ -1,182 +1,290 @@
 package renter
 
 import (
+	"bytes"
 	"errors"
-	"fmt"
-	"io"
 	"os"
-	"strings"
 	"sync/atomic"
 	"time"
 
-	"github.com/NebulousLabs/Sia/crypto"
 	"github.com/NebulousLabs/Sia/modules"
-	"github.com/NebulousLabs/Sia/modules/renter/contractor"
+	"github.com/NebulousLabs/Sia/types"
 )
 
 var (
 	errInsufficientHosts  = errors.New("insufficient hosts to recover file")
 	errInsufficientPieces = errors.New("couldn't fetch enough pieces to recover data")
 )
 
-// A fetcher fetches pieces from a host. This interface exists to facilitate
-// easy testing.
-type fetcher interface {
-	// pieces returns the set of pieces corresponding to a given chunk.
-	pieces(chunk uint64) []pieceData
+// chunkDownload tracks the progress of a chunk.
+type chunkDownload struct {
+	// Map of workers who have attempted downloads for this chunk, from the
+	// worker id to the index of the piece they are trying to download.
+	workerAttempts map[types.FileContractID]uint64
 
-	// fetch returns the data specified by piece metadata.
-	fetch(pieceData) ([]byte, error)
+	// List of completed pieces, mapping from the index of the piece to the
+	// data.
+	completedPieces map[uint64][]byte
```

Comment:
I guess you are saying that the chunk map should be built up here instead of where it is done later in the code. I'll think about it, but that feels like over-optimization to me.